fuse_optimizers/src/fixed_lag_smoother.cpp

```diff
@@ -495,20 +496,100 @@ void FixedLagSmoother::transactionCallback(
+ * - FAILURE, USER_FAILURE     -> ERROR (default)
+ *
+ * @param[in] termination_type The optimization termination type
+ * @return The diagnostic status with the level and message corresponding to the optimization termination type
+ */
+diagnostic_msgs::DiagnosticStatus terminationTypeToDiagnosticStatus(const ceres::TerminationType termination_type)
+{
+  switch (termination_type)
   {
-    status.message = "FixedLagSmoother " + status.message;
+    case ceres::TerminationType::CONVERGENCE:
+    case ceres::TerminationType::USER_SUCCESS:
+      return makeDiagnosticStatus(diagnostic_msgs::DiagnosticStatus::OK, "Optimization converged");
+    case ceres::TerminationType::NO_CONVERGENCE:
+      return makeDiagnosticStatus(diagnostic_msgs::DiagnosticStatus::WARN, "Optimization didn't converged");
+    default:
+      return makeDiagnosticStatus(diagnostic_msgs::DiagnosticStatus::ERROR, "Optimization failed");
   }
+}
+
+void FixedLagSmoother::setDiagnostics(diagnostic_updater::DiagnosticStatusWrapper& status)
+{
+  Optimizer::setDiagnostics(status);
 
-  status.add("Started", started_);
+  // Load std::atomic<bool> flag that indicates whether the optimizer has started or not
+  const bool started = started_;
+
+  status.add("Started", started);
   {
     std::lock_guard<std::mutex> lock(pending_transactions_mutex_);
     status.add("Pending Transactions", pending_transactions_.size());
   }
+
+  if (started)
+  {
+    // Add some optimization summary report fields to the diagnostics status if the optimizer has started
+    if (optimization_mutex_.try_lock())
+    {
+      const auto summary = summary_;
+      optimization_mutex_.unlock();
+
+      if (summary.total_time_in_seconds >= 0.0)
+      {
+        status.add("Optimization Termination Type", ceres::TerminationTypeToString(summary.termination_type));
+        status.add("Optimization Total Time [s]", summary.total_time_in_seconds);
+        status.add("Optimization Iterations", summary.iterations.size());
+        status.add("Initial Cost", summary.initial_cost);
+        status.add("Final Cost", summary.final_cost);
+
+        status.mergeSummary(terminationTypeToDiagnosticStatus(summary.termination_type));
+      }
+    }
+    else
+    {
+      status.add("Optimization", "Running");
+    }
+
+    // Add time since the last optimization request time. This is useful to detect if no transactions are received for
+    // too long
+    if (optimization_requested_mutex_.try_lock())
+    {
+      const auto optimization_deadline = optimization_deadline_;
+      optimization_requested_mutex_.unlock();
+
+      if (!optimization_deadline.isZero())
+      {
+        const auto optimization_request_time = optimization_deadline - params_.optimization_period;
+        const auto time_since_last_optimization_request = ros::Time::now() - optimization_request_time;
+        status.add("Time Since Last Optimization Request [s]", time_since_last_optimization_request.toSec());
+      }
+    }
```

Comment:
And similar here:

```suggestion
    auto optimization_deadline = decltype(optimization_deadline_)();
    {
      std::unique_lock<std::mutex> lock(optimization_requested_mutex_, std::try_to_lock);
      if(lock)
      {
        optimization_deadline = optimization_deadline_;
      }
    }
    if (!optimization_deadline.isZero())
    {
      const auto optimization_request_time = optimization_deadline - params_.optimization_period;
      const auto time_since_last_optimization_request = ros::Time::now() - optimization_request_time;
      status.add("Time Since Last Optimization Request [s]", time_since_last_optimization_request.toSec());
    }
```

Reply:
Done